### DISCUSSION TITLE
Error on delete with zero records

### DISCUSSION THREAD

**Discussion by rgaino** [prisma/prisma#21682] [Q&A] [answered]
### Question

I want to understand the rationale behind throwing an exception when calling delete on a where clause that returns 0 records.

```
An operation failed because it depends on one or more records that were required but not found. Record to delete does not exist.
```

Shouldn't it simply complete successfully affecting 0 records? This forces us to perform a count operation first and only proceed if it's > 0. I've noticed this behaviour on Postgres.

Thank you

**Comment by ludralph** [comment, marked as answer]
Hi @rgaino 👋 

Thank you for raising this question.

The behavior you're experiencing is indeed how Prisma currently operates. When you attempt to delete a record that doesn't exist using the `delete` method, Prisma throws an exception. This is because the `delete` method is designed to operate on a single record and expects that record to exist. If the record doesn't exist, it throws an error to indicate that the operation couldn't be completed as expected.

This behavior can be useful in scenarios where the absence of a record is an exceptional circumstance and should halt the execution of your code. However, it can indeed lead to additional database operations if you need to check for the existence of a record before attempting to delete it.

There have been discussions and open issues in the Prisma community about this behavior. For example, in this GitHub issue  [#10142](https://github.com/prisma/prisma/issues/10142), a user suggested adding `updateOrThrow()` and `deleteOrThrow()` methods to Prisma Client, which would throw an exception if the record to update or delete doesn't exist. In GitHub issue [#9460](https://github.com/prisma/prisma/issues/9460), a user suggested adding a `deleteIfExists` method, which would delete a record if it exists and do nothing if it doesn't, without throwing an error.

As a [workaround](https://github.com/prisma/prisma/issues/4072#issuecomment-1256254284), you can use the deleteMany method instead of delete. The deleteMany method doesn't throw an error if no records are found that match the where clause. It simply returns a count of the records deleted, which would be 0 if no matching records were found.
Here's an example:
```js
const deleteResult = await prisma.types.deleteMany({
  where: {
    bigserial: "non-existent id",
  },
});
```
In this case, `deleteResult.count` would be 0 if no records were found that match the where clause.

Please note that this is a workaround and might not be suitable for all use cases. Always consider the implications of using deleteMany instead of delete, especially in scenarios where deleting multiple records could have unintended side effects.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by rgaino** [comment]
Thanks for your quick and details answer @ludralph . I indeed used `deleteMany` after I typed this question.

**Reply by stefanatanasovmk** [reply]
Me too, it was very well explained answer. 

Do you have any update about the suggested `updateOrThrow` and `deleteOrThrow` features?
